lib/semantics/mod-file.cc

```diff
@@ -362,12 +347,53 @@ std::ostream &PutLower(std::ostream &os, const std::string &str) {
   return os;
 }
 
+// Write the module file at path, prepending header. Return false on error.
+static bool WriteFile(const std::string &path, std::string &&contents) {
+  std::fstream stream;
+  auto header{GetHeader(contents)};
+  auto size{GetFileSize(path)};
+  if (size == header.size() + 1 + contents.size()) {
+    // file exists and has the right size, check the contents
+    stream.open(path, std::ios::in | std::ios::out);
+    if (FileContentsMatch(stream, header, contents)) {
+      return true;
+    }
+    stream.seekp(0);
+  } else {
+    stream.open(path, std::ios::out);
+  }
+  stream << header << '\n' << contents;
+  stream.close();
+  return !stream.fail();
+}
+
+// Return true if the stream matches what we would write for the mod file.
+static bool FileContentsMatch(std::fstream &stream, const std::string &header,
+    const std::string &contents) {
+  char c;
+  for (std::size_t i{0}; i < header.size(); ++i) {
+    if (!stream.get(c) || c != header[i]) {
+      return false;
+    }
+  }
+  if (!stream.get(c) || c != '\n') {
+    return false;
+  }
+  for (std::size_t i{0}; i < contents.size(); ++i) {
+    if (!stream.get(c) || c != contents[i]) {
+      return false;
+    }
+  }
+  return true;
+}
+
 // Compute a simple hash of the contents of a module file and
 // return it as a string of hex digits.
 // This uses the Fowler-Noll-Vo hash function.
-std::string CheckSum(const std::string &str) {
+template<typename Iter> static std::string CheckSum(Iter begin, Iter end) {
   std::uint64_t hash{0xcbf29ce484222325ull};
-  for (char c : str) {
+  for (auto it{begin}; it != end; ++it) {
+    char c{*it};
     hash ^= c & 0xff;
```

Comment:
Do you need this mask if you use "unsigned char" as the type in the line above?

Reply:
braced initialization with `unsigned char` would fail to compile if the signedness of default char was signed.  Using the explicit masking is most portable.